From f33827af160b90ac26952b084dbdfec529dd82aa Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:01 -0400
Subject: [scsi] qla2xxx: Fix undefined behavior in call to snprintf()

Message-id: <1408391817-6389-14-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87923
O-Subject: [PATCH RHEL 7.1 e-stor 13/69] qla2xxx: Fix undefined behavior in call to snprintf().
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Joe Carnuccio <joe.carnuccio@qlogic.com>

scsi commit id ad5fa025d8fdc850f4ebf34d523a99739aae3e33

Signed-off-by: Joe Carnuccio <joe.carnuccio@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_attr.c b/drivers/scsi/qla2xxx/qla_attr.c
index 5f174b8..dd0b72f 100644
--- a/drivers/scsi/qla2xxx/qla_attr.c
+++ b/drivers/scsi/qla2xxx/qla_attr.c
@@ -917,8 +917,8 @@ qla2x00_serial_num_show(struct device *dev, struct device_attribute *attr,
   return snprintf(buf, PAGE_SIZE, "%s\n",
       vha->hw->mr.serial_num);
  } else if (IS_FWI2_CAPABLE(ha)) {
-  qla2xxx_get_vpd_field(vha, "SN", buf, PAGE_SIZE);
-  return snprintf(buf, PAGE_SIZE, "%s\n", buf);
+  qla2xxx_get_vpd_field(vha, "SN", buf, PAGE_SIZE - 1);
+  return strlen(strcat(buf, "\n"));
  }
 
  sn = ((ha->serial0 & 0x1f) << 16) | (ha->serial2 << 8) | ha->serial1;
-- 
1.7.1